_sips/minutes/2018-09-24-sip-minutes.md

```diff
@@ -0,0 +1,218 @@
+abstraction of code that depends on a notion/representation of a context,
+like Scala 3's compiler) and points out that implicit function types can
+replace the reader monad, even though it's about 10x faster than the reader monad
+is. A comprehensive explanation of what implicit function types can do can be
+found in [Olivier's Blainvillain talk at
+ScalaDays, Berlin 2018](https://slideslive.com/38908156/applications-of-implicit-function-types).
+**Martin** thinks that implicit function types should be seen as the
+canonical way of doing scope injection, which gives you a lot of
+expressivity, to which **Sebastien** adds that what Martin means by canonical
+scope injection doesn't necessarily correspond with the way people do normal
+scope injection, because in normal scope injection you can't refer to
+identifier or parent scopes. **Martin** clarifies that for him comonadic
+abstraction are the classical way of scope injection in which we inject
+things into an environment, hence the use of canonical scope injection when
+referring to implicit function types which allow you to do the same. His
+definition of scope injection comes more from a typing rules perspective
+rather than the lexical point of view. **Martin** agrees that if there is a
+name clash with implicit function types there is a problem indeed.
+
+#### [Proposal to add Dependent Function Types to the Language](https://dotty.epfl.ch/docs/reference/dependent-function-types.html)
+ ([YouTube time: 43’11’’ - 44’40’’](https://youtu.be/tEb4UF6RJrM?t=2591))
+
+**Martin** mentions that dependent function types is the last big addition to Scala's type checker. The reason why they are added is because Scala has dependent methods and there is a need for dependent functions (the same rationale has been doing with regards to implicit methods and implicit function types). It's an obvious win because dependent function types allow us to abstract over the idea of implicit methods in functions, so the more we can do the better. Initially he was afraid of the feature because he thought it violated this Scala principle that in the end anything is an instance of a class in some way and it turned out that a new encoding of dependent function types made this initial argument moot. Dependent function types are now encoded as implicit function types with type refinements, so this way it doesn't violate that principle.
+**Adriaan** mentions that the last missing bit is polymorphic function types
+and Martin agrees and says that they are looking into that, but maybe not for
+Scala 3.0 (Guillaume Martres is pushing for polymorphic function types).
+
+#### [Proposal to add Trait Parameters to the Language](https://dotty.epfl.ch/docs/reference/trait-parameters.html)
+
+([YouTube time: 44’42’’ - end ](https://youtu.be/tEb4UF6RJrM?t=2682)) 
+
+**Martin** describes trait parameters and says that they subsume a large number of use
+cases of early initializers. They were not added to Scala 2 because of
```

Comment:
"to Scala 2" is misleading here I think. I think a better rendering of the idea is "to Scala from the start". We might still add them to Scala 2 (hopefully we will). The point Martin made was that in initially designing Scala, the only reason he did not include trait parameters was the uncertainty [...].